File: 2019/4/solution.js
const fs = require('fs');
const path = require('path');

const input = fs.readFileSync(path.join(__dirname, 'input.txt'), 'utf-8', (err, data) => {
  if (err) {
    console.log(err)
  } else {
    return data;
  }
}).split('-').map(Number);

const adjacentSameDigitsRegex = /(.)\1+/g;

class Password {
  constructor(password) {
    this.password = password.toString();
    this.part1 = true;
    this.part2 = true;
    this.check();
  }

  check() {
    if (this.password.length != 6) {
      this.part1 = false;
      this.part2 = false;
      return;
    }
    const passwordArray = this.password.split('').map(Number);
    for (let i = 1; i < passwordArray.length; i++) {
      if (passwordArray[i] < passwordArray[i - 1]) {
        this.part1 = false;
        this.part2 = false;
        return;
      }
    }
    const matches = this.password.match(adjacentSameDigitsRegex);
    if (matches === null) {
      this.part1 = false;
      this.part2 = false;
      return;
    }
    this.part2 = matches.some((a) => a.length === 2);
  }
}

let part1 = 0;
let part2 = 0;
for (let i = input[0]; i <= input[1]; i++) {
  const password = new Password(i);
  if (password.part1) {
    part1++;
  }
  if (password.part2) {
    part2++;
  }
}

console.log(part1);
console.log(part2);

File: 2019/4/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	puzzleInput := parseInput(string(data))
	fmt.Println(part1(puzzleInput))
	fmt.Println(part2(puzzleInput))
}

func parseInput(data string) [2]int {
	parts := strings.Split(data, "-")
	var nums [2]int
	if n, err := strconv.Atoi(parts[0]); err != nil {
		panic(err)
	} else {
		nums[0] = n
	}
	if n, err := strconv.Atoi(parts[1]); err != nil {
		panic(err)
	} else {
		nums[1] = n
	}
	return nums
}

func isValid(n int, stricter bool) bool {
	a, b := n/10, n%10
	twins := make(map[int]int)
	for a > 0 {
		c, d := a/10, a%10
		if d > b {
			return false
		}
		if d == b {
			twins[d]++
		}
		a, b = c, d
	}
	if !stricter {
		return len(twins) > 0
	}
	for _, n := range twins {
		if n == 1 {
			return true
		}
	}
	return false
}

func part1(puzzleInput [2]int) int {
	res := 0
	for n := puzzleInput[0]; n <= puzzleInput[1]; n++ {
		if isValid(n, false) {
			res++
		}
	}
	return res
}

func part2(puzzleInput [2]int) int {
	res := 0
	for n := puzzleInput[0]; n <= puzzleInput[1]; n++ {
		if isValid(n, true) {
			res++
		}
	}
	return res
}
